app/version.go

```diff
@@ -1,40 +1,89 @@
 package app
 
-import "github.com/urfave/cli/v2"
+import (
+	"fmt"
+	"runtime/debug"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/urfave/cli/v2"
+	"golang.org/x/mod/semver"
+)
 
 const (
-	DefaultVersion = "v0.13.0"
+	pseudoVersionMinLen        = len("vX.0.0-yyyymmddhhmmss-abcdefabcdef")
+	pseudoVersionCommitInfoLen = len("yyyymmddhhmmss-abcdefabcdef")
 )
 
 var (
-	BuildDate string
-	Commit    string
-	Version   string
+	MakeVersion string
 )
 
+type BuildInfo struct {
+	Commit     string
+	CommitTime time.Time
+	Version    string
+}
+
 func NewVersionCommand() (CommandOut, error) {
 	return CommandOut{Command: &cli.Command{
 		Name:    "version",
 		Usage:   "Version information",
 		Aliases: []string{"v"},
 		Action: func(c *cli.Context) error {
-			return PrintObj(&struct {
-				BuildDate string
-				Commit    string
-				Version   string
-			}{
-				BuildDate: BuildDate,
-				Commit:    Commit,
-				Version:   getVersion(),
-			})
+			return PrintObj(NewBuildInfo())
 		},
 	}}, nil
 }
 
-func getVersion() string {
-	version := Version
-	if len(version) == 0 {
-		version = DefaultVersion
+func NewBuildInfo() BuildInfo {
+	var info BuildInfo
+
+	di, ok := debug.ReadBuildInfo()
+	if !ok {
+		return info
 	}
-	return version
+
+	if semver.IsValid(di.Main.Version) {
+		info.Version = strings.Split(semver.Canonical(di.Main.Version), "-")[0]
+
+		// See https://go.dev/ref/mod#pseudo-versions for more info on the expected string format
+		// when the binary is compiled with go install. We always expect to hit this path if
+		// di.Main.Version is a valid semver.
+		if len(di.Main.Version) >= pseudoVersionMinLen {
+			commitInfoStart := len(di.Main.Version) - pseudoVersionCommitInfoLen
+			split := strings.Split(di.Main.Version[commitInfoStart:], "-")
+			info.Commit = split[1]
+
+			at, err := time.Parse("20060102150405", split[0])
+			if err == nil {
+				info.CommitTime = at
+			}
+
+			return info
```

Comment:
I'm not following what this block is for? Is it if the version is a commit sha and not a semver?

Reply:
Its when the version is filled out by `(devel)`. Go does this because they are strict about tagging versions only if it is reproducible, meaning the tag can't change. Thats also why their pseudo versions include timestamps and commit hashes.

I re-structured the code and added comments so it is more clear.